security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,423 @@
+          callback(endorsementError, null);
+        } else {
+          let hasher = crypto.createHash('sha256');
+          hasher.update(endorsementKey);
+          this._registrationId = (base32Encode(hasher.digest(), 'RFC4648').toLowerCase()).replace(/=/g, '');
+          callback(null, this._registrationId);
+        }
+      }.bind(this));
+    }
   }
 
-  getStorageRootKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  private _createPersistentPrimary(pki: PersistentKeyInformation, callback: (err: Error, resultPublicKey: TPMT_PUBLIC) => void): void {
+    this._tpm.allowErrors().ReadPublic(pki.handle, (resp: tss.ReadPublicResponse) => {
+      let rc = this._tpm.getLastResponseCode();
+      debug('ReadPublic(' + pki.name + ') returned ' + TPM_RC[rc] +  (rc === TPM_RC.SUCCESS ? '; PUB: ' + resp.outPublic.toString() : ''));
+      if (rc !== TPM_RC.SUCCESS) {
+        this._tpm.withSession(tss.NullPwSession).CreatePrimary(pki.hierarchy, new tss.TPMS_SENSITIVE_CREATE(), pki.template, null, null, (resp: tss.CreatePrimaryResponse) => {
+          debug('CreatePrimary(' + pki.name + ') returned ' + TPM_RC[this._tpm.getLastResponseCode()] + '; pub size: ' + (resp.outPublic.unique as tss.TPM2B_PUBLIC_KEY_RSA).buffer.length);
+          this._tpm.withSession(tss.NullPwSession).EvictControl(tss.Owner, resp.handle, pki.handle, () => {
+            debug('EvictControl(0x' + resp.handle.handle.toString(16) + ', 0x' + pki.handle.handle.toString(16) + ') returned ' + TPM_RC[this._tpm.getLastResponseCode()]);
+            this._tpm.FlushContext(resp.handle, () => {
+              debug('FlushContext(TRANSIENT_' + pki.name + ') returned ' + TPM_RC[this._tpm.getLastResponseCode()]);
+              callback(null, resp.outPublic);
+            });
+          });
+        });
+      } else {
+        callback(null, resp.outPublic);
+      }
+    });
   }
 
-  signWithIdentity(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  private _signData(dataToSign: Buffer, callback: (err: Error, signedData: Buffer) => void): void {
+
+    let idKeyHashAlg: TPM_ALG_ID = (<tss.TPMS_SCHEME_HMAC>(<tss.TPMS_KEYEDHASH_PARMS>this._idKeyPub.parameters).scheme).hashAlg;
+
+    this._tpm.GetCapability(tss.TPM_CAP.TPM_PROPERTIES, TPM_PT.INPUT_BUFFER, 1, (caps: tss.GetCapabilityResponse) => {
+      let props = <tss.TPML_TAGGED_TPM_PROPERTY>caps.capabilityData;
+      if (props.tpmProperty.length !== 1 || props.tpmProperty[0].property !== TPM_PT.INPUT_BUFFER) {
+        callback(new errors.DeviceRegistrationFailedError('Unexpected result of TPM2_GetCapability(TPM_PT.INPUT_BUFFER)'), null);
+      } else {
+        let maxInputBuffer: number = props.tpmProperty[0].value;
+        if (dataToSign.length <= maxInputBuffer) {
+          this._tpm.withSession(tss.NullPwSession).HMAC(idKeyPersHandle, dataToSign, idKeyHashAlg, (signature: Buffer) => {
+            callback(null, signature);
```

Comment:
>callback(null, signature) [](start = 12, length = 25)

<groan> This code implies that there's a bunch of functions on _tpm that _cannot_ fail.  I find this really hard to believe.

Reply:
You and me both.  Will be requesting changes to MSR to make api's more nodish.

---
In reply to: [159311721](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r159311721) [](ancestors = 159311721)